### PR TITLE
Make BrowserAction actually WERK

### PR DESCRIPTION
**Pull request by strseb**
This is my current branch i demoed.... 
i tried to make this easier to review by making small commit but this is still quite big. SORRY ❤️

### CODE REVIEW

File: src/background/requestHandler.js [hunk 1]
```diff
@@ -18,72 +22,90 @@ export class RequestHandler extends Component {
...
   /**
    *
-   * @param { RequestDetails } requestInfo
+   * @param { proxy.RequestDetails } requestInfo
    * @returns
    */
   async interceptRequests(requestInfo) {
-    let { url, originUrl } = requestInfo;
+    let { documentUrl } = requestInfo;
+    if (documentUrl != "https://mullvad.net/en/check") {
```
Comment (lesleyjanenorton) [inline src/background/requestHandler.js:89]: 🤔

File: src/background/requestHandler.js [hunk 2]
```diff
@@ -18,72 +22,90 @@ export class RequestHandler extends Component {
...
+      debugger;
+    }
 
-    for (let urlString of [url, originUrl]) {
```
Comment (lesleyjanenorton) [inline src/background/requestHandler.js:80]: Curious: have you seen any requests where `documentUrl` is different than `originUrl` ? I think we'll need to keep `url` here for the situations where someone is on `aSiteWithoutProxyConfig` with requests going to `aSiteThatDoesHaveAProxyConfig`, which we'd want to reroute according to the rule.

File: src/background/requestHandler.js [hunk 3]
```diff
@@ -18,72 +22,90 @@ export class RequestHandler extends Component {
...
+    for (let urlString of [documentUrl]) {
       if (urlString) {
-        const parsedHostname = await Utils.getFormattedHostname(urlString);
-        const siteContext = self.siteContexts.get(parsedHostname);
-        if (siteContext) {
-          log("Intercepting web request. Redirect to local proxy.");
-          return [...siteContext.proxyInfo];
+        const parsedHostname = Utils.getFormattedHostname(urlString);
+        const proxyInfo = this.proxyMap.get(parsedHostname);
+        console.log([parsedHostname, proxyInfo]);
```
Comment (lesleyjanenorton) [inline src/background/requestHandler.js:106]: we probably don't want to log this :)

**Review by lesleyjanenorton — APPROVED** [review]
R+ with commentary

File: src/background/requestHandler.js (hunk 1 above)
Comment (strseb) [reply src/background/requestHandler.js:89]: OOPS sorry that must go

File: src/background/requestHandler.js [hunk 4]
```diff
@@ -94,6 +116,68 @@ export class RequestHandler extends Component {
...
+ * @param {VPNState} vpnState
+ * @returns {Map<string, Array<browser.proxy.proxyInfo>>}
+ */
+export const resolveSiteContext = (siteContexts, vpnState) => {
+  /** @type {Map<string, Array<browser.proxy.proxyInfo>>} */
+  const result = new Map();
+  const localProxy = vpnState.loophole
+    ? [ProxyUtils.parseProxy(vpnState.loophole)]
+    : [];
+  siteContexts.forEach((ctx, origin) => {
```
[the remark this answers is not in the document]
Comment (strseb) [reply src/background/requestHandler.js:151]: I tried to pull this out into a pure function so our "where does this site go?" becomes trivial to test. 
So in theory this function i would envision this function to check the VPN state and react to that :)

File: src/background/requestHandler.js (hunk 3 above)
Comment (strseb) [reply src/background/requestHandler.js:106]: trueee

File: src/background/requestHandler.js (hunk 2 above)
Comment (strseb) [reply src/background/requestHandler.js:80]: If you do a 3rdparty-request, yes. 
Just go to whatismyip.com and you will see requests with originURL 
```
Array [ "c.amazon-adsystem.com", undefined ]
requestHandler.js:101:17
Array [ "fonts.googleapis.com", undefined ]
requestHandler.js:101:17
Array [ "fonts.gstatic.com", undefined ]
requestHandler.js:101:17
Array [ "config.aps.amazon-adsystem.com", undefined ]
requestHandler.js:101:17
Array [ "tpc.googlesyndication.com", undefined ]
requestHandler.js:101:17
Array [ "secure.cdn.fastclick.net", undefined ]
requestHandler.js:101:17
Array [ "tags.crwdcntrl.net", undefined ]
```

> on aSiteWithoutProxyConfig with requests going to aSiteThatDoesHaveAProxyConfig,

We should not do that imho. 

``` 
evil.com (rule: Berlin)
whatIsMyIP.com (rule: None and the user is Connected to Brazil. )
```
and evil.com does `fetch(whatIsMyIP.com)` evil.com has just learned another IP the user is using right now. 


Firefox "total cookie protection" is partitions storage by top the level document origin, so i think for consistency we should do that too :)

**Pull request merged**